content/blog/next-level-iac-breakpoint-debugging/index.md

```diff
@@ -0,0 +1,317 @@
+
+Outputs:
+    cdnHostname   : output<string>
+    cdnURL        : output<string>
+    originHostname: output<string>
+    originURL     : output<string>
+
+Resources:
+    + 8 to create
+
+Do you want to perform this update? yes
+Updating (dev)
+
+View in Browser (Ctrl+O): https://app.pulumi.com/troy-pulumi-corp/good-morning/dev/updates/1
+
+     Type                                   Name                 Status
+ +   pulumi:pulumi:Stack                    good-morning-dev       created (200s)
+ +   ├─ aws:s3:Bucket                       bucket               created (1s)
+ +   ├─ aws:s3:BucketPublicAccessBlock      public-access-block  created (0.57s)
+ +   ├─ aws:s3:BucketOwnershipControls      ownership-controls   created (1s)
+ +   ├─ aws:cloudfront:Distribution         cdn                  created (196s)
+ +   └─ synced-folder:index:S3BucketFolder  bucket-folder        created (0.41s)
+ +      ├─ aws:s3:BucketObject              error.html           created (0.46s)
+ +      └─ aws:s3:BucketObject              index.html           created (0.46s)
+
+Outputs:
+    cdnHostname   : "##############.cloudfront.net"
+    cdnURL        : "https://##############.cloudfront.net"
+    originHostname: "bucket-#######.s3-website-us-west-2.amazonaws.com"
+    originURL     : "http://bucket-#######.s3-website-us-west-2.amazonaws.com"
+
+Resources:
+    + 8 created
+
+Duration: 3m21s
+```
+
+Great! You should be able to navigate to the website using the URL from the `cdnURL` output value.
+
+```shell
+$ open $(pulumi stack output cdnURL)
+```
+
+If all looks good, let's move on to the next step by opening your project directory in VS Code:
+
+```shell
+$ code .
+```
+
+### Step 2: Create a debug wrapper script 
+
+VS Code can run an arbitrary command when you hit `F5`. Typically this would be a one-liner that runs your language runtime like `node foo.js` or `python foo.py`. In our case, we need to run the `pulumi` CLI and need to do more than one step. The easiest way to do this is to create a small wrapper script and call that from VS Code.
+
+In the root of the directory, create a file called `pulumi-debug.sh`:
+
+```shell
+#!/bin/bash
+
+pulumi login
+pulumi stack select dev
+pulumi up -f
+```
+
+{{% notes type="info" %}}
+Let's walk through these lines to explain what's going on here:
+
+* `pulumi login` to set up credentials for the pulumi backend
+* `pulumi stack select dev` to run this against our `dev` stack
+* `pulumi up -f` to run our program udpate. The -f flag skips the preview phase and also skips the need to interactively confirm that we want to run the update.
+{{% /notes %}}
+
+Open the VS Code terminal panel and first set it to be executable, then try running the script: 
+
+```shell
+$ chmod +x pulumi-debug.sh
+$ ./pulumi-debug.sh
+```
+
+Ok now let's setup VS Code to run that in debug mode when we hit `F5`!
+
+### Step 3: Configure VS Code `launch.json` and `task.json`
```

Comment:
I think what is missing in all of this is we show "how to set up and use a debuguger", but we don't show a sample of something that is broken and using a debugger to help us fix it. I think the thinking here is like people always code errors into their programs, it will end up executing and failing at provisioning. for something like CloudFormation its a long roll back process. for TF, I don't know what happens. But in Pulumi you can debug the program with a debugger and isolate the error.

Reply:
Showing a simple error that will normally creep up and how to find it is probably the piece that is missing

Reply:
Good point! I originally wanted to write it this way but felt the complexity and length of the post would be too much. I went ahead and added that in anyway, in the current revision.